lgc/util/PassManager.cpp

```diff
@@ -108,25 +133,90 @@ static AnalysisID getPassIdFromName(StringRef passName) {
+// Create a LegacyPassManagerImpl
 lgc::LegacyPassManager *lgc::LegacyPassManager::Create() {
+  return new LegacyPassManagerImpl;
+}
+
+// =====================================================================================================================
+// Create a PassManagerImpl
+lgc::PassManager *lgc::PassManager::Create() {
   return new PassManagerImpl;
 }
 
 // =====================================================================================================================
-PassManagerImpl::PassManagerImpl() : LegacyPassManager() {
+LegacyPassManagerImpl::LegacyPassManagerImpl() : LegacyPassManager() {
   if (!cl::DumpCfgAfter.empty())
     m_dumpCfgAfter = getPassIdFromName(cl::DumpCfgAfter);
 
   m_jumpThreading = getPassIdFromName("jump-threading");
   m_printModule = getPassIdFromName("print-module");
 }
 
+// =====================================================================================================================
+PassManagerImpl::PassManagerImpl() : PassManager(), instrVerify(getLgcOuts()) {
+  if (!cl::DumpCfgAfter.empty()) {
+    // TODO: We need to support m_dumpCfgAfter in a way that is similar to what
+    // is done in the add() function of the legacy pass manager. We can use the
+    // getPassNameForClassName function of PassInstrumentationCallbacks to get
+    // a pass ID from its name but we need to provide a mechanism to register
+    // custom passes there too.
+    report_fatal_error("-dump-cfg-after cannot be used with the new pass manager yet.");
+  }
+  // Setup custom instrumentation callbacks and register LLVM's default module
+  // analyses to the analysis manager.
+  registerCallbacks();
+  if (cl::VerifyIr)
+    instrVerify.registerCallbacks(instrCallbacks);
+  PassBuilder passBuilder(nullptr, PipelineTuningOptions(), None, &instrCallbacks);
+  passBuilder.registerModuleAnalyses(analysisManager);
+}
+
+// =====================================================================================================================
+void PassManagerImpl::run(Module &module) {
+  ModulePassManager::run(module, analysisManager);
+}
+
+// =====================================================================================================================
+void PassManagerImpl::registerCallbacks() {
+  // Before running a pass, we increment the pass index if it exists, and we
+  // dump the pass name if needed.
+  auto beforePass = [this](StringRef passName, Any ir) {
+    if (passName != PrintModulePass::name() && m_passIndex) {
+      unsigned passIndex = (*m_passIndex)++;
+      if (cl::DumpPassName)
+        LLPC_OUTS("Pass[" << passIndex << "] = " << passName << "\n");
+    }
+  };
+  instrCallbacks.registerBeforeSkippedPassCallback(beforePass);
+  instrCallbacks.registerBeforeNonSkippedPassCallback(beforePass);
+
+  instrCallbacks.registerShouldRunOptionalPassCallback([this](StringRef passName, Any ir) {
+    // Do not add any passes after calling stop().
+    if (m_stopped)
+      return false;
```

Comment:
If the "stop" mechanism is not needed in the new pass manager, let's not port it.